Implement indirect draw validation

teoxoy opened the pull request:
**Connections**
Closes https://github.com/gfx-rs/wgpu/issues/2431.

**Description**
Implements indirect draw validation by batching draw calls and injecting a compute pass prior to the render pass that will validate the indirect args.

NOTES

- This PR doesn't cover https://github.com/gfx-rs/wgpu/issues/2471
- MDI is ~not~ covered (see https://github.com/gfx-rs/wgpu/pull/7140#discussion_r1974123889).
- We can probably use the same bind group for both draw and dispatch validation passes but I haven't looked into it yet.

**Testing**
Added new tests, probably not enough, there are lots of possible permutations. I looked at branch coverage and most branches are tested at least.

The commits are self contained and so the PR doesn't need to be squashed.

teoxoy committed:
implement indirect draw validation

cwfitzgerald commented on tests/tests/draw_indirect.rs:
Might consider having a test for weird overlapping indirect calls. Like a call at offset 0, one at 4, one at 8.

teoxoy replied:
I added a new test for this.

cwfitzgerald commented on wgpu-core/src/command/bundle.rs:
```diff
@@ -103,7 +103,7 @@ use crate::{
 use arrayvec::ArrayVec;
 
-use std::{borrow::Cow, mem::size_of, num::NonZeroU32, ops::Range, sync::Arc};
+use std::{borrow::Cow, num::NonZeroU32, ops::Range, sync::Arc};
```
When updating, should make sure these are set to `core`/`alloc`

cwfitzgerald commented on wgpu-core/src/command/bundle.rs:
```diff
@@ -774,29 +774,38 @@ fn multi_draw_indirect(
+    #[cfg(feature = "indirect-validation")]
+    let buffer_uses = wgt::BufferUses::STORAGE_READ_ONLY;
+    #[cfg(not(feature = "indirect-validation"))]
+    let buffer_uses = wgt::BufferUses::INDIRECT;
```
Could we use a normal if statement for this? DCE/constant prop will take care of it

cwfitzgerald commented on wgpu-core/src/command/bundle.rs:
```diff
@@ -1018,19 +1036,28 @@ impl RenderBundle {
+                        *_instance_limit,
+                    )?;
+                    #[cfg(not(feature = "indirect-validation"))]
+                    let (buffer, offset) = (buffer.try_raw(snatch_guard)?, *offset);
```
Thought: `cfg_if` might end up easier to understand here. I'm not an amazingly big fan of this... compile time polymorphism is easy to accidentally break, especially if `indirect-validation` is almost always enabled.

cwfitzgerald commented on wgpu-core/src/indirect_draw_validation.rs:
```diff
@@ -0,0 +1,1080 @@
+    track::TrackerIndex,
+    FastHashMap,
+};
+use std::{
```
Mind the std

cwfitzgerald commented on wgpu-core/src/indirect_draw_validation.rs:
```diff
@@ -0,0 +1,1080 @@
+///
+/// - 65535 [`wgt::DrawIndirectArgs`] / [`MetadataEntry`]
+/// - 52428 [`wgt::DrawIndexedIndirectArgs`]
+const BUFFER_SIZE: wgt::BufferSize = unsafe { wgt::BufferSize::new_unchecked(1048560) };
```
```suggestion
const BUFFER_SIZE: wgt::BufferSize = unsafe { wgt::BufferSize::new_unchecked(1_048_560) };
```

It'd be so nice to have this not be unsafe, but Option::unwrap is only const in 1.83

cwfitzgerald commented on wgpu-core/src/indirect_draw_validation.rs:
```diff
@@ -0,0 +1,1080 @@
+        "failed |= first_instance != 0u;"
+    };
+
+    format!(
```
Could we move this to a new file - what I've done in the tests is use `str::replace` to replace `{{indirect_first_instance_check}}` in the file source with the check we want. Alternatively, couldn't a pipeline override used with:

```
if disable_indirect_first_instance {
    failed |= first_instance != 0u;
}
```

then DCE should do its thing in the backend.

cwfitzgerald commented on wgpu-core/src/indirect_draw_validation.rs:
```diff
@@ -0,0 +1,1080 @@
+    };
+    let pipeline =
+        unsafe { device.create_compute_pipeline(&pipeline_desc) }.map_err(|err| match err {
+            hal::PipelineError::Device(error) => {
+                CreateComputePipelineError::Device(DeviceError::from_hal(error))
+            }
+            hal::PipelineError::Linkage(_stages, msg) => CreateComputePipelineError::Internal(msg),
+            hal::PipelineError::EntryPoint(_stage) => CreateComputePipelineError::Internal(
+                crate::device::ENTRYPOINT_FAILURE_ERROR.to_string(),
+            ),
+            hal::PipelineError::PipelineConstants(_, error) => {
+                CreateComputePipelineError::PipelineConstants(error)
+            }
```
Thought: could these be in some kind of `From` implementation?

teoxoy replied:
I don't think we should, device errors in `hal::PipelineError::Device` need to go through `device.handle_hal_error()` unless there is no device yet (in this case). A `From` impl would be easy to misuse.

I think at some point we should coalesce the `Linkage` and `EntryPoint` variants though.

cwfitzgerald reviewed (changes requested):
I think my only overarching comment would be - I wonder if we can somehow internalize the `cfgs` in some way so that we don't need to have them all over the codebase. 

~~I also worry about the non-indirect buffer pass starting to fail to compile and us not noticing~~ WebGL2 compiles without it.

teoxoy committed:
Replace the `indirect-validation` feature with `InstanceFlags::VALIDATION_INDIRECT_CALL`.

With the only caveat that device creation will now panic if the `wgsl` feature is not enabled, `InstanceFlags::VALIDATION_INDIRECT_CALL` is set and the device supports `DownlevelFlags::INDIRECT_EXECUTION`.

teoxoy commented:
I ended up removing the `indirect-validation` feature, see last commit.

cwfitzgerald commented on wgpu/Cargo.toml:
```diff
@@ -175,11 +175,7 @@ static_assertions.workspace = true
-    "renderdoc",
-    "indirect-validation",
-] }
+wgpu-core = { workspace = true, features = ["raw-window-handle", "renderdoc"] }
```
Having this be disabled in `wgpu` feels wrong - there definitely should be some way of turning this on

teoxoy replied:
Will enable `wgsl` unconditionally on native as discussed.

cwfitzgerald reviewed (changes requested):
Discussed some feedback in call

teoxoy committed:
add `indirect_buffer_offsets` test

cwfitzgerald commented on wgpu-core/src/command/bundle.rs:
```diff
@@ -1029,19 +1045,29 @@ impl RenderBundle {
-                    indexed: true,
+                    indexed,
+
+                    vertex_or_index_limit: _vertex_or_index_limit,
+                    instance_limit: _instance_limit,
```
These shouldn't need to be `_` prefixed anymore

cwfitzgerald commented on wgpu-core/src/command/render.rs:
```diff
@@ -2517,6 +2546,36 @@ fn multi_draw_indirect(
         ),
     );
 
+    let (indirect_raw, offset) = if count == 1 && state.device.indirect_draw_validation.is_some() {
```
Would there be significant lift to dealing with MDI by doing these calls in a loop?

teoxoy replied:
I was thinking we would need a fastpath for MDI but maybe we don't (or can add it later). I will add the loop.

teoxoy replied:
Ah, no, this doesn't work; unless we call `state.raw_encoder.draw_indirect/draw_indexed_indirect` in the loop as well since the destination buffer might change between calls to `IndirectDrawValidationBatcher.add()`.

teoxoy replied:
Or we could keep track of the destination buffer and only call `draw_indirect/draw_indexed_indirect` once it changes or at the end of the loop.

teoxoy replied:
I pushed this as a separate commit.

cwfitzgerald commented on wgpu-core/src/device/queue.rs:
```diff
@@ -267,6 +267,9 @@ pub(crate) struct EncoderInFlight {
     inner: crate::command::CommandEncoder,
     pub(crate) trackers: Tracker,
     pub(crate) temp_resources: Vec<TempResource>,
+    #[allow(unused)] // We only need to keep these resources alive.
+    indirect_draw_validation_resources:
+        crate::indirect_draw_validation::IndirectDrawValidationResources,
```
Couldn't this be:

```suggestion
    // We only need to keep these resources alive.
    _indirect_draw_validation_resources:
        crate::indirect_draw_validation::IndirectDrawValidationResources,
```

cwfitzgerald commented on wgpu-core/src/device/resource.rs:
```diff
@@ -235,11 +238,13 @@ impl Device {
-            .flags
-            .contains(wgt::DownlevelFlags::INDIRECT_EXECUTION)
-        {
+        let enable_indirect_validation = instance_flags
+            .contains(wgt::InstanceFlags::VALIDATION_INDIRECT_CALL)
+            && downlevel
+                .flags
+                .contains(wgt::DownlevelFlags::INDIRECT_EXECUTION);
```
Thought: can there be a device that has indirect execution, but not compute shaders?

teoxoy replied:
Only our GL backend sets those 2 features differently (and metal, but it always supports compute shaders):

https://github.com/gfx-rs/wgpu/blob/a734e86041beb383e8d330b5230e9a05f2bcd288/wgpu-hal/src/gles/adapter.rs#L312-L313

https://github.com/gfx-rs/wgpu/blob/a734e86041beb383e8d330b5230e9a05f2bcd288/wgpu-hal/src/gles/adapter.rs#L380-L381

The thing is, I don't think it's right that we check for `GL_ARB_multi_draw_indirect` as it doesn't provide indirect dispatches.

teoxoy replied:
I will remove `GL_ARB_multi_draw_indirect` and document that `INDIRECT_EXECUTION` implies presence of `COMPUTE_SHADERS`.

teoxoy replied:
Actually `GL_ARB_compute_shader` enables indirect dispatches.

teoxoy replied:
I pushed this as a separate commit.

cwfitzgerald commented on wgpu-core/src/indirect_validation.rs:
Thought: Feel free to boot this to another PR if it would be a pain: I feel like this could be better organized as an indirect_validation module containing two submodules: `compute` and `draw`, with common code in the parent module.

teoxoy replied:
Agree, I tried to not touch the old code as part of this PR but it should be fine as a separate commit.

cwfitzgerald commented on wgpu-core/src/lock/observing.rs:
```diff
@@ -80,6 +80,11 @@ impl<T> Mutex<T> {
         }
     }
+
+    #[allow(dead_code)]
```
Unneeded dead code?

cwfitzgerald commented on wgpu-core/src/lock/rank.rs:
```diff
@@ -148,6 +148,8 @@ define_lock_ranks! {
     rank BLAS_BUILT_INDEX "Blas::built_index" followed by { }
     rank TLAS_BUILT_INDEX "Tlas::built_index" followed by { }
     rank TLAS_DEPENDENCIES "Tlas::dependencies" followed by { }
+    #[allow(dead_code)]
```
Same here?

cwfitzgerald commented on wgpu-core/src/lock/vanilla.rs:
```diff
@@ -30,6 +30,11 @@ impl<T> Mutex<T> {
         MutexGuard(self.0.lock())
     }
+
+    #[allow(dead_code)]
```
And here

cwfitzgerald commented on wgpu-core/src/resource.rs:
```diff
@@ -787,8 +794,8 @@ pub struct DestroyedBuffer {
     label: String,
     bind_groups: WeakVec<BindGroup>,
-    #[cfg(feature = "indirect-validation")]
     raw_indirect_validation_bind_group: Option<Box<dyn hal::DynBindGroup>>,
+    raw_indirect_draw_validation_bind_group: Option<Box<dyn hal::DynBindGroup>>,
```
Similar nit to module naming, I feel like we should have naming parity here - draw_validation and compute_validation

teoxoy replied:
I ended up combining those 2 in the new module.

cwfitzgerald reviewed (changes requested):
Tests look _fabulous_, have some comments and a note about structure and naming, but I like the current way of dealing with shaders, at least for now.

teoxoy committed:
guarantee `DownlevelFlags::COMPUTE_SHADERS` when `DownlevelFlags::INDIRECT_EXECUTION` is enabled

teoxoy committed:
validate multi draw indirect calls

teoxoy committed:
move indirect validation code to its own folder

rename items appropriately
internalize indirect buffer bind groups

cwfitzgerald commented on wgpu-core/src/indirect_validation/draw.rs:
```diff
@@ -0,0 +1,984 @@
+
+fn create_bind_group_layout(
+    device: &dyn hal::DynDevice,
+    read_only: bool,
+    has_dynamic_offset: bool,
+    min_binding_size: wgt::BufferSize,
```
Could we pass a `wgt::BindingType` instead of three separate arguments

teoxoy replied:
`wgt::BindingType` is an enum though, any of its variants could be passed in.

cwfitzgerald commented on wgpu-core/src/indirect_validation/draw.rs:
```diff
@@ -0,0 +1,984 @@
+        }
+        buffer_index_set.clear();
+
+        for index in batches
+            .values()
+            .map(|batch| batch.metadata_resource_index)
+            .filter(|index| buffer_index_set.insert(*index))
```
Could we do these as early-continues in the for loop - as we're iterating already, I think it would read more clearly

teoxoy replied:
Related: I was worried about those manual `.clear()` calls. I added 2 new data structures to get rid of the possibility of forgetting to do things in the right order.

Let me know what you think!

cwfitzgerald commented on wgpu-core/src/indirect_validation/validate_draw.wgsl:
```diff
@@ -0,0 +1,82 @@
+    let vertex_or_index_count = src[src_base_offset + 0];
+
+    {
+        let can_overflow = ((metadata.dst_offset >> 30) & 1u) == 1u;
```
this would be a bit clearer using `extractBits` - if you don't want to use it for performance reasons, maybe a free function?

teoxoy replied:
The backends add some validation code for `extractBits` to cap the offset and count. I'll add a function.

cwfitzgerald commented on wgpu-types/src/instance.rs:
```diff
@@ -98,6 +98,11 @@ bitflags::bitflags! {
         /// When `Self::from_env()` is used takes value from `WGPU_GPU_BASED_VALIDATION` environment variable.
         const GPU_BASED_VALIDATION = 1 << 4;
+
+        /// Validate indirect buffer content prior to issuing indirect draws/dispatches.
+        ///
+        /// When `Self::from_env()` is used takes value from `WGPU_VALIDATION_INDIRECT_CALL` environment variable.
+        const VALIDATION_INDIRECT_CALL = 1 << 5;
```
I have thoughts on how this isn't the right place to put this - but that is _not_ this PR's to solve :)

cwfitzgerald replied:
Filed https://github.com/gfx-rs/wgpu/issues/7261

cwfitzgerald reviewed (changes requested):
Basically there, some style comments.

teoxoy committed:
[validate_draw] refactor out an `is_bit_set` function

teoxoy committed:
abstract away usage of scratch space for buffer barriers and unique index iterators

cwfitzgerald commented on wgpu-core/src/indirect_validation/utils.rs:
```diff
@@ -0,0 +1,84 @@
+    pub(crate) fn new(scratch: &'a mut BufferBarrierScratch<'_>) -> Self {
+        // change lifetime of buffer reference, this is safe since `scratch` is empty,
+        // it was either just created or it has been cleared on `BufferBarriers::drop`
+        let scratch = unsafe {
+            core::mem::transmute::<&'a mut BufferBarrierScratch<'_>, &'a mut BufferBarrierScratch<'b>>(
+                scratch,
+            )
+        };
```
Could we talk about this on a call when you're back? This don't make a lot of sense to me why we need this compared to normal vectors

teoxoy replied:
Resolving since we talked about it and long term we should use a non-default allocator to do this.

cwfitzgerald reviewed (approved):
I have one main question, but I don't think we should block on it.

This is awesome!
